### PR TITLE
Introduce an rpm-controlled per-build directory

### PR DESCRIPTION
**Pull request by pmatilai**
In our ancient wacko setup, %_builddir is shared by all your packages, under which a package may create %buildsubdir - if it uses %setup that is. In other words, there's no safe and sane place for rpm to place per-build material. Introduce a new intermediate directory between the two, created in a newly added (internal) %builddir build step, to give rpm that place. This opens up all manner of opportunities, to be explored in later commits.

A new build-time macro %pkgbuilddir is added for the absolute path of this new directory, but in addition we override %_builddir to the same value for maximum compatibility with existing specs - a LOT of packages reference %{_builddir} for all sorts of (mostly bad) reasons and we don't want to deal with the carnage that would follow from breaking it. A further complication here is that defining %_builddir (along with sources etc) to current working directory is a common configuration (used eg by fedpkg and its variants) that we don't want to break either. So upon entry, we grab the pre-existing %_builddir and define %_top_builddir from it, which we can then base the %pkgbuilddir on, while preserving compatibility with both of the above scenarios.

Fixes: #2078

### CODE REVIEW

**Comment by pmatilai** [comment]
FWIW, the %{NAME}-%{VERSION}-PKG name is not really intended as final, I used that just to have it stand out from the logs. Better ideas welcome.

**Comment by kloczek** [comment]
> Do I understand correctly that the `BUILDROOT` dir was replaced by `%{_builddir}/%{_target_cpu}-%{_target_os}-root`? The `%{_builddir}` is the right move IMHO, but what is the advantage of `%{_target_cpu}-%{_target_os}-root` over `BUILDROOT`, especially when e.g. `SPECPARTS` stays the same.

I'm using below patch
```patch
--- a/tests/rpmspec.at
+++ b/tests/rpmspec.at
@@ -333,8 +333,8 @@
   exit $STATUS
 fi
 cd 'hello-1.0'
-rm -rf '/build/BUILD/hello-1.0-SPECPARTS'
-/usr/bin/mkdir -p '/build/BUILD/hello-1.0-SPECPARTS'
+rm -rf '/build/BUILD/hello-1.0/SPECPARTS'
+/usr/bin/mkdir -p '/build/BUILD/hello-1.0/SPECPARTS'
 /usr/bin/chmod -Rf a+rX,u+w,g-w,o-w .
 echo "Patch #0 (hello-1.0-modernize.patch):"
 /usr/bin/patch --no-backup-if-mismatch -f -p1 -b --suffix .modernize --fuzz=0 < /build/SOURCES/hello-1.0-modernize.patch
--- a/macros.in
+++ b/macros.in
@@ -271,7 +271,7 @@
 %buildroot             %{_buildrootdir}/%{NAME}-%{VERSION}-%{RELEASE}.%{_arch}

 #      Path for spec file snippets generated during build
-%specpartsdir %{_builddir}/%{buildsubdir}-SPECPARTS
+%specpartsdir %{_builddir}/%{buildsubdir}/SPECPARTS

 #      Directory where temporaray files can be created.
 %_tmppath              %{_var}/tmp
--- a/build/build.c
+++ b/build/build.c
@@ -195,14 +195,8 @@
        fprintf(fp, "cd '%s'\n", buildSubdir);

     if (what == RPMBUILD_RMBUILD) {
-       if (rpmMacroIsDefined(spec->macros, "specpartsdir")) {
-           char * buf = rpmExpand("%{specpartsdir}", NULL);
-           fprintf(fp, "rm -rf '%s'\n", buf);
-           free(buf);
-       }
        if (buildSubdir[0] != '\0')
-           fprintf(fp, "rm -rf '%s' '%s.gemspec'\n",
-                   buildSubdir, buildSubdir);
+           fprintf(fp, "rm -rf '%s'\n", buildSubdir);
     } else if (sb != NULL)
        fprintf(fp, "%s", sb);

```
After that I've tested that whit almost 5k of my src.rpm and found that only few packages are failing because of that change.
All those packages are python modules and fail s result of MISSING parts in pyprojects toml.

Because SPECPARTS **exposed those issue** IMO instead evacuating that directory outside source tree it should be INTENTIONALLY put in that location.

In meantime for all those cases I've opened issue tickets and ~half already have been sorted out.

Example fail on building `uc-micro-py` python module:
```console
+ /usr/bin/python3 -sBm build -w --no-isolation
* Getting build dependencies for wheel...
No `packages` or `py_modules` configuration, performing automatic discovery.
`flat-layout` detected -- analysing .
discovered packages -- ['uc_micro', 'SPECPARTS ', 'uc_micro.categories', 'uc_micro.properties', 'uc_micro.categories.Cc', 'uc_micro.categories.Cf', 'uc_micro.categories.P', 'uc_micro.categories.Z', 'uc_micro.properties.Any']
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
    main()
  File "/usr/lib/python3.8/site-packages/pyproject_hooks/_in_process/_in_process.py", line 335, in main
    json_out['return_val'] = hook(**hook_input['kwargs'])
  File "/usr/lib/python3.8/site-packages/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel
    return hook(config_settings)
  File "/usr/lib/python3.8/site-packages/setuptools/build_meta.py", line 325, in get_requires_for_build_wheel
    return self._get_build_requires(config_settings, requirements=['wheel'])
  File "/usr/lib/python3.8/site-packages/setuptools/build_meta.py", line 295, in _get_build_requires
    self.run_setup()
  File "/usr/lib/python3.8/site-packages/setuptools/build_meta.py", line 311, in run_setup
    exec(code, locals())
  File "<string>", line 1, in <module>
  File "/usr/lib/python3.8/site-packages/setuptools/__init__.py", line 103, in setup
    return distutils.core.setup(**attrs)
  File "/usr/lib/python3.8/site-packages/setuptools/_distutils/core.py", line 159, in setup
    dist.parse_config_files()
  File "/usr/lib/python3.8/site-packages/setuptools/dist.py", line 627, in parse_config_files
    pyprojecttoml.apply_configuration(self, filename, ignore_option_errors)
  File "/usr/lib/python3.8/site-packages/setuptools/config/pyprojecttoml.py", line 66, in apply_configuration
    config = read_configuration(filepath, True, ignore_option_errors, dist)
  File "/usr/lib/python3.8/site-packages/setuptools/config/pyprojecttoml.py", line 131, in read_configuration
    return expand_configuration(asdict, root_dir, ignore_option_errors, dist)
  File "/usr/lib/python3.8/site-packages/setuptools/config/pyprojecttoml.py", line 156, in expand_configuration
    return _ConfigExpander(config, root_dir, ignore_option_errors, dist).expand()
  File "/usr/lib/python3.8/site-packages/setuptools/config/pyprojecttoml.py", line 204, in expand
    self._expand_all_dynamic(dist, package_dir)
  File "/usr/lib/python3.8/site-packages/setuptools/config/pyprojecttoml.py", line 249, in _expand_all_dynamic
    version=self._obtain_version(dist, package_dir),
  File "/usr/lib/python3.8/site-packages/setuptools/config/pyprojecttoml.py", line 298, in _obtain_version
    return _expand.version(self._obtain(dist, "version", package_dir))
  File "/usr/lib/python3.8/site-packages/setuptools/config/pyprojecttoml.py", line 287, in _obtain
    return self._expand_directive(
  File "/usr/lib/python3.8/site-packages/setuptools/config/pyprojecttoml.py", line 281, in _expand_directive
    return _expand.read_attr(directive["attr"], package_dir, root_dir)
  File "/usr/lib/python3.8/site-packages/setuptools/config/expand.py", line 188, in read_attr
    _parent_path, path, module_name = _find_module(module_name, package_dir, root_dir)
  File "/usr/lib/python3.8/site-packages/setuptools/config/expand.py", line 229, in _find_module
    if package_dir:
  File "/usr/lib/python3.8/site-packages/setuptools/config/expand.py", line 459, in __len__
    return len(self._target())
  File "/usr/lib/python3.8/site-packages/setuptools/config/expand.py", line 452, in _target
    self._value = self._obtain()
  File "/usr/lib/python3.8/site-packages/setuptools/config/expand.py", line 422, in _get_package_dir
    self()
  File "/usr/lib/python3.8/site-packages/setuptools/config/expand.py", line 412, in __call__
    self._dist.set_defaults(name=False)  # Skip name, we can still be parsing
  File "/usr/lib/python3.8/site-packages/setuptools/discovery.py", line 353, in __call__
    self._analyse_package_layout(ignore_ext_modules)
  File "/usr/lib/python3.8/site-packages/setuptools/discovery.py", line 387, in _analyse_package_layout
    or self._analyse_flat_layout()
  File "/usr/lib/python3.8/site-packages/setuptools/discovery.py", line 444, in _analyse_flat_layout
    return self._analyse_flat_packages() or self._analyse_flat_modules()
  File "/usr/lib/python3.8/site-packages/setuptools/discovery.py", line 450, in _analyse_flat_packages
    self._ensure_no_accidental_inclusion(top_level, "packages")
  File "/usr/lib/python3.8/site-packages/setuptools/discovery.py", line 480, in _ensure_no_accidental_inclusion
    raise PackageDiscoveryError(cleandoc(msg))
setuptools.errors.PackageDiscoveryError: Multiple top-level packages discovered in a flat-layout: ['uc_micro', 'SPECPARTS '].

To avoid accidental inclusion of unwanted files or directories,
setuptools will not proceed with this build.

If you are trying to create a single distribution with multiple packages
on purpose, you should not rely on automatic discovery.
Instead, consider the following options:

1. set up custom discovery (`find` directive with `include` or `exclude`)
2. use a `src-layout`
3. explicitly set `py_modules` or `packages` with a list of names

To find more information, look for "package discovery" on setuptools docs.

ERROR Backend subprocess exited when trying to invoke get_requires_for_build_wheel
```

**Comment by kloczek** [comment]
Just made grep across all Fedora spec files
```console
[tkloczko@pers-jacek SPECS.fedora]$ grep SPECPARTS  *
python-bidict.spec:# SPECPARTS dir in %%_builddir/%%buildsubdir is leaking to setuptools package
python-bidict.spec:rm -rf SPECPARTS
python-cairocffi.spec:- Drop rmdir SPECPARTS
python-chaospy.spec:rm -rf SPECPARTS
python-linkify-it-py.spec:- Stop removing the SPECPARTS directory, it doesn't exist
python-linkify-it-py.spec:- Remove empty SPECPARTS directory that breaks build
python-mechanize.spec:rm -rf SPECPARTS
python-neo.spec:# remove rpm's SPECPARTS file
python-neo.spec:rm -rf SPECPARTS
python-nihtest.spec:# Work around issue with package discovery due to SPECPARTS dir
python-nose2.spec:rm -rf SPECPARTS
python-numpoly.spec:rm -rf SPECPARTS
python-quantities.spec:# Work around confusion with SPECPARTS directory looking like a package to
python-quantities.spec:rm -rf SPECPARTS
python-uc-micro-py.spec:rm -rf SPECPARTS
qcad.spec:rm -rf ../*-SPECPARTS
rpm.spec:- Use mkdir -p for creating SPECPARTS dir
traceroute.spec:%make_build CFLAGS="$RPM_OPT_FLAGS" LDFLAGS="$RPM_LD_FLAGS" SKIPDIRS="${RPM_SPECPARTS_DIR##*/}"
```
Few of those cases already have been sorted out by source tree maintainers.

Another question is .. is it really worth to keep SPECPARTS extension in rpm? 🤔 
So far not been able to find even SINGLE case of use that.
IMO that whole SPECPARTS created so far more issues (and by this wasted more man/hours) than saved any time on packaging.

**Comment by kloczek** [comment]
Just found in inbox that another maintainer sorted out issue related to create directory in python module source tree https://github.com/PyGithub/PyGithub/pull/2894. As you see fixing that it is not rocket science ..

**Comment by pmatilai** [comment]
> Do I understand correctly that the `BUILDROOT` dir was replaced by `%{_builddir}/%{_target_cpu}-%{_target_os}-root`? The `%{_builddir}` is the right move IMHO, but what is the advantage of `%{_target_cpu}-%{_target_os}-root` over `BUILDROOT`, especially when e.g. `SPECPARTS` stays the same.

These details are nowhere near fully thought through, and are certainly debatable and open/subject to change. In my initial patch buildroot was always BUILDROOT, but then experimenting with the vpath build stuff and your own mention about potentially needing multiple build directories made me think perhaps it should match the build directory name, and those are commonly named by the target tuple. 

All that is of course irrelevant within a single rpmbuild run, but down that road far in the horizon one could imagine unpacking the source once, and then executing just the build steps from other hosts using a read-only shared mount of the source. Or (cross-) building on the same host for multiple targets at once. Of course in such a scenario, SPECPARTS would have to be similarly per-arch-os too, it's a valid point.

Wild visions aside, think of the (buildroot) rename more like shaking a tree a bit to see what falls out rather than something decided.

**Comment by kloczek** [comment]
> @kloczek , this is not about SPECPARTS although if you _look_ at the PR, that's one of the issues that gets solved by this. See #2532 for the background.

On first look it looks like it is `setuptools` issue. .. like in other python modules cases (which I've mention here).

**Comment by voxik** [comment]
> These details are nowhere near fully thought through, and are certainly debatable and open/subject to change. In my initial patch buildroot was always BUILDROOT, but then experimenting with the vpath build stuff and your own mention about potentially needing multiple build directories made me think perhaps it should match the build directory name, and those are commonly named by the target tuple.

While I think multiple directories for vpath builds are useful, if `BUILDROOT` is still the image of what will be installed on the resulting system, then my current thinking is that there should be only one destination, where everything should be parallel installable if possible. But maybe you had the `curl` example with `RemovePathPostfixes` on your mind which could be legitimate reason.

**Comment by pmatilai** [comment]
Multiple buildroots could be useful for the the "variant builds" case, but the arch-os naming doesn't help with *that* at all, the potential benefits I see are more far fetched. But, it's useful to shake things a bit, it doesn't *have* to be BUILDROOT just because we've had such a thing in the past.

**Comment by dmnks** [comment]
Not a fan of the `-root` suffix. Now that we have the option, why not take full advantage of directories and have a tree like this:

```
BUILD/
├── bar-1.0
│   ├── ROOT
│   │   └── x86_64-linux
│   │   └── [...]
│   └── SPECPARTS
└── foo-1.0
    ├── ROOT
    │   └── x86_64-linux
    │   └── [...]
    └── SPECPARTS
```

File: build/build.c [hunk 1]
```diff
@@ -303,6 +300,24 @@ static rpmRC doCheckBuildRequires(rpmts ts, rpmSpec spec, int test)
     return rc;
 }
 
+static rpmRC doBuildDir(rpmSpec spec, int test, StringBuf *sbp)
+{
+    char *doDir = rstrscat(NULL,
+			   "rm -rf ", spec->buildDir, "\n",
```
Comment (dmnks) [inline build/build.c:306]: We probably want to quote the argument to rm -rf, just in case 🥶 😄

Comment (pmatilai) [reply build/build.c:306]: Oops, yep

**Comment by pmatilai** [comment]
> Not a fan of the `-PKG` and `-root` suffixes. 

The -PKG is ugly for sure, but *some* differentation from the traditional %setup created directory seems necessary to drive the point across, and make logs easier to look at. If you just have 'foo-1.0/foo-1.0/' its easy to mix up etc. NEVRA might be more reasonable, and for --rebuild mode it could actually be randomized.

> Now that we have the option, why not take full advantage of directories and have a tree like this:
> [...]
> While I like the singular `ROOT` better, maybe `ROOTS` would be more consistent.

For now we only have one buildroot though. I guess I'll just go back to the original BUILDROOT inside the new build directory, at least people will know what that thing is :sweat_smile:

**Comment by voxik** [comment]
> > Not a fan of the `-PKG` and `-root` suffixes.
> 
> If you just have 'foo-1.0/foo-1.0/' its easy to mix up etc.

I don't mind this path. But if the path actually was 'foo-1.0/SOURCE/' that could mitigate your concern. And yes, there is #2859 always expanding the tarball into name-version directory, so if the expansion was instead always into `SRC` that would not be so bad IMHO.

But arguably, quite some projects place their real sources into something like `src` directory, so then there could be patch like 'foo-1.0/SOURCE/src' 🙈

**Comment by pmatilai** [comment]
The thing is (well, one of the things) is that, foo-1.0/foo-1.0 can mask a bug or packaging error. The ugly -PKG suffix was absolutely necessary for seeing which path a thing is actually trying to use when developing this PR, and will be equally necessary for packagers trying to troubleshoot something more exotic.

So it doesn't need to be -PKG but it needs to be something that is easily distinguishable from something %setup normally does.

File: build/build.c (hunk 1 above)
Comment (pmatilai) [reply build/build.c:306]: Quotes added now.

**Comment by pmatilai** [comment]
Merged this into a single commit at least for now as the changing paths clashed in maddening ways in the test-suite on rebases, making simple changes too hard to bother with. 

Buildroot is just BUILDROOT now. I would've used name-version-release.arch for the per-build directory, but this turns out to break a buildid no-recompute test which assumes that a buildid remains the same regardless of the release. So it's name-version-arch now, for the better or worse.

**Comment by pmatilai** [comment]
Other ponderings include the per-build directory macro name, should it be just %builddir without the underscore (instead of %pkgbuilddir)

**Comment by dmnks** [comment]
One thing to keep in mind here is that we'll be getting rid of a shared BUILDROOT. I've always wondered what the purpose of that (or the shared `%_topdir` workspace in general) was, but I can think of one use case:

You wish to deploy a common set of packages and/or configuration (a *suite*) to a number of hosts in your organization. You want that suite to be managed and updated "atomically" as a whole, and ideally distributed as a single binary RPM without additional dependencies. This is where the shared RPM workspace comes into play! You simply build those individual packages locally one by one, and finally build the "meta" package that just packages the resulting BUILDROOT.

Basically, this is what container images are in the modern world. I don't think the above use case is very common (or real at all) but I can imagine that being one of the motivations behind the shared workspace back then.

**Comment by dmnks** [comment]
Which makes me think - couldn't the shared BUILDROOT be useful for actually building container images? I'm not sure about the advantages over just grabbing pre-built RPMs to compose the final root filesystem tree, but it does seem like you'd save a number of redundant steps if you were building those packages yourself.

**Comment by dmnks** [comment]
To summarize my above comments a bit, from a higher-level perspective:

In the context of the shared `%_topdir`, an RPM package doesn't necessarily have to correspond to a single program or piece of software. It's a way to distribute a "snapshot" or "sub-tree" of the root filesystem. In that context, a shared workspace makes a bit more sense. With this PR, we'd be removing that piece of flexibility. But again, that's probably what we really want, or at least it's not something we really care for.

**Comment by dmnks** [comment]
Thinking more about it, the shared BUILDROOT use case might actually be impossible to achieve because of the fact that RPM checks for unpackaged files in there when building a single package (see the recent discussions around excludes).

**Comment by pmatilai** [comment]
Eh, except that we create our scriptlets, including the one that creates the builddir, in %_tmppath so we there's a tiny :egg: - :chicken: matter there. Oh well, maybe later.

**Comment by pmatilai** [comment]
After (way too) much mulling over it: BUILDROOT and SPECPARTS names are now hardcoded in the source, neither of these should be user configurable really but were so for various historical reasons, such as not having the guaranteed per-build directory we're adding in this very PR. Removed the corresponding macros from the main macros file too.

File: build/build.c (hunk 1 above)
Comment (dmnks) [reply build/build.c:306]: Seems like this has reoccurred :smile:

**Comment by dmnks** [comment]
Speaking of name bikeshedding,

> Other ponderings include the per-build directory macro name, should it be just %builddir without the underscore (instead of %pkgbuilddir)

This has gotten lost in the above chatter, but I think it's worth considering. It's a macro name that we'll be living with from now on and somehow `%builddir` feels more natural than `%pkgbuilddir`. To me, that is.

OTOH, the "pkg" prefix perhaps emphasizes that it's a *per-package* build directory as opposed to some common one (which is `%_top_builddir` now with the patch). So dunno. Maybe I'm overthinking it...

**Comment by dmnks** [comment]
As for the paths, I think they're fine now, no need to mull over those more, indeed :smile:

**Comment by dmnks** [comment]
Lastly, the `docs/manual/dynamic_specs.md` files needs updating as it mentions the old SPECPARTS path.

**Comment by pmatilai** [comment]
> Update 2: Is there an actual need for a SPEC to ever use the %pkgbuilddir (or %builddir for that matter) macro? If not, why not make it at least "private" with an underscore?

Oh, I actually mulled about %builddir without the underscore as the user-oriented name for this. I too kinda liked that more, I think I ended up with the pkg-version to make minimize confusion with %_builddir. OTOH, like you said we already have %buildroot and %builddir would seem more consistent with that. I could quite easily be convinced either way :smile: 

> Update 3: Oh... you actually mentioned that "a LOT of packages reference %_builddir for all sorts of (mostly bad) reasons" in the commit message. In that case, why introduce yet another such macro at all?

Primarily because %_builddir is ambiguous, you don't really know whether it's the bad old or good new. It still points to the potentially shared %{_topdir}/BUILD by default, whereas %pkgbuilddir (or %builddir) is only defined during the actual build (which is also the only time that directory exists).

Note that the "mostly bad" stops being bad with this change - this is now the legit place for packages to use for their needs during the build.

**Comment by pmatilai** [comment]
Good point on the docs, that dynamic stuff certainly needs updating, but there could be other things too. Like, explaining the new layout somehow. Not that packagers *need* to know but it doesn't exactly hurt either. 

It's possible this could use a test or two as well...

**Comment by pmatilai** [comment]
Ended up renaming %pkgbuilddir to %builddir as per your and my gut feelings - we don't have "pkg" anywhere at all in our macros so introducing it here would seem a bit strange somehow.
Updated the docs a bit + added couple of tests to demonstrate the new layout and behavior.

As a side-note, SPECPARTS behaves a little funny here because it's only created from %setup (because there was no other place to do it before), but could be moved to %builddir script.

Ehh, that made me notice there's an ambiguity in the naming now: the script and the directory that the script creates are both called %builddir. Maybe the script needs to be called %mkbuilddir instead.

**Comment by pmatilai** [comment]
Renamed the builddir creation script to %mkbuilddir (along with the enum) in the last push to disambiguate the two.

File: build/parsePreamble.c [hunk 2]
```diff
@@ -1267,28 +1267,31 @@ int parsePreamble(rpmSpec spec, int initialPackage)
...
-	char *buildRoot = rpmGetPath(spec->buildRoot, NULL);
-	free(spec->buildRoot);
-	spec->buildRoot = buildRoot;
-	rpmPushMacro(spec->macros, "buildroot", NULL, spec->buildRoot, RMIL_SPEC);
-	if (*buildRoot == '\0') {
-	    rpmlog(RPMLOG_ERR, _("%%{buildroot} couldn't be empty\n"));
-	    goto exit;
+	/* Grab the top builddir on first entry as we'll override _builddir */
+	if (!rpmMacroIsDefined(spec->macros, "_top_builddir")) {
+	    rpmPushMacro(spec->macros, "_top_builddir", NULL, "%{_builddir}",
```
Comment (mlschroe) [inline build/parsePreamble.c:1277]: I may be confused, but don't you need to expand %{_buildir} first?

File: build/parsePreamble.c [hunk 3]
```diff
@@ -1267,28 +1267,31 @@ int parsePreamble(rpmSpec spec, int initialPackage)
...
+	    rpmPushMacro(spec->macros, "_builddir", NULL, spec->buildDir, RMIL_SPEC);
+
+	    /* A user-oriented, unambiguous name for the thing */
+	    rpmPushMacro(spec->macros, "builddir", NULL, spec->buildDir, RMIL_SPEC);
+	    spec->buildRoot = rpmGetPath(spec->buildDir, "/BUILDROOT", NULL);
+	    rpmPushMacro(spec->macros, "buildroot", NULL, spec->buildRoot, RMIL_SPEC
+);
+	    char *specparts = rpmGetPath(spec->buildDir, "/SPECPARTS", NULL);
+	    rpmPushMacro(spec->macros, "specpartsdir", NULL, specparts, RMIL_SPEC);
+	    free(specparts);
```
Comment (mlschroe) [inline build/parsePreamble.c:1294]: Maybe you should use rpmPushMacroFlags and add the RPMMACRO_LITERAL flag

**Comment by dmnks** [comment]
Seems like the test-suite changes have accidentally spilled into the preceding commit `Print informative messages to stderr, not stdout on --target build` :smile:

File: build/parsePreamble.c (hunk 2 above)
Comment (dmnks) [reply build/parsePreamble.c:1277]: This seems wrong, indeed; the value is (recursively) expanded through `%{_top_builddir}` in the `rpmExpand()` call right away, *but* that call only happens if `spec->buildDir` is unset.

Comment (pmatilai) [reply build/parsePreamble.c:1277]: Oh indeed. The logic for setting %builddir does work, but %_top_builddir will end up pointing to %builddir instead which is not the intention.

I'll fixup and make all these macros use RPMMACRO_LITERAL, that's the right thing here really.

**Comment by pmatilai** [comment]
Thirteenth time the charm, eh? :smile: 
Should be fixed up now, and thanks guys for spotting this stuff.

File: build/build.c (hunk 1 above)
Comment (dmnks) [reply build/build.c:311]: Don't forget about this one :smile:

Comment (pmatilai) [reply build/build.c:311]: Aagh :laughing:

Comment (pmatilai) [reply build/build.c]: Fixed now, and thanks for the reminder. Had totally forgotten it...

File: build/parsePreamble.c [hunk 4]
```diff
@@ -1267,28 +1267,38 @@ int parsePreamble(rpmSpec spec, int initialPackage)
...
+	    /* Grab top builddir on first entry as we'll override _builddir */
+	    if (!rpmMacroIsDefined(spec->macros, "_top_builddir")) {
+		char *top_builddir = rpmExpand("%{_builddir}", NULL);
+		rpmPushMacroFlags(spec->macros, "_top_builddir", NULL,
+				top_builddir, RMIL_GLOBAL, RPMMACRO_LITERAL);
+		free(top_builddir);
+	    }
+
+	    /* Using release here causes a buildid no-recompute test to fail */
+	    spec->buildDir = rpmExpand("%{_top_builddir}/%{NAME}-%{VERSION}-%{_arch}", NULL);
```
Comment (Conan-Kudo) [inline build/parsePreamble.c:1285]: Can we please not include architecture? That can leak into builds and cause problems. That's why I had to make an adjustment in redhat-rpm-config for `%_vpath_builddir`: https://src.fedoraproject.org/rpms/redhat-rpm-config/c/e0cfcc0fc76a7642faabb25c5e348d6a1314ace2

Comment (pmatilai) [reply build/parsePreamble.c:1285]: Oh and FWIW, I would love to get the full build paths somehow abstracted out anyhow. I don't know how much is really doable in that space, but on systems that support this kind of stuff, builds would run temp directory inside their build directory and writes limited to their own build directory, bind-mounted so that the build thinks its running in /build directory instead of /home/whatever/somewhere path etc.

**Comment by lantw44** [comment]
Is it possible to shorten the `<package>-<version>-build` part from a spec file?

I package Guix in Fedora Copr. To run Guix tests in the rpmbuild %check step, the full path of the working directory cannot be longer than 29 bytes. The default `/builddir/build/BUILD/guix-1.4.0` used by mock is already too long, so I have to change the default `test-tmp` directory used by Guix tests to `t` to save a few bytes in order to run tests.

Now the new default `/builddir/build/BUILD/guix-1.4.0-build/guix-1.4.0` is even longer than `/builddir/build/BUILD/guix-1.4.0`, so the `test-tmp` → `t` trick can't work. Is it possible to tell rpmbuild to use a shorter path than `guix-1.4.0-build`, or I should just `mktemp -d` and tell Guix to run tests in the temporary directory?

**Comment by pmatilai** [comment]
@lantw44: A closed PR is not the place for questions. Please open a [discussion](https://github.com/rpm-software-management/rpm/discussions) instead.
